tsl/src/continuous_aggs/create.c

```diff
@@ -1283,31 +1284,34 @@ cagg_validate_query(const Query *query, const bool finalized, const char *cagg_s
 					op = (OpExpr *) join->quals;
 					rte = list_nth(query->rtable, ((RangeTblRef *) join->larg)->rtindex - 1);
 					rte_other = list_nth(query->rtable, ((RangeTblRef *) join->rarg)->rtindex - 1);
+					if (rte->subquery != NULL || rte_other->subquery != NULL)
+						ereport(ERROR,
+								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								 errmsg("invalid continuous aggregate view"),
+								 errdetail("sub-queries are not supported in FROM clause")));
+					RangeTblEntry *jrte = rt_fetch(join->rtindex, query->rtable);
+					if (jrte->joinaliasvars == NIL)
+						ereport(ERROR,
+								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								 errmsg("invalid continuous aggregate view")));
 				}
 			}
 		}
 
-		/*
-		 * Cagg with joins does not support hierarchical caggs in from clause.
-		 */
-		if (rte->relkind == RELKIND_VIEW || rte_other->relkind == RELKIND_VIEW)
-			ereport(ERROR,
-					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					 errmsg("joins for hierarchical continuous aggregates are not supported")));
-
 		/*
 		 * Error out if there is aynthing else than one normal table and one hypertable
-		 * in the from clause, e.g. sub-query.
+		 * in the from clause, e.g. sub-query, lateral, two hypertables, etc.
 		 */
-		if (((rte->relkind != RELKIND_RELATION && rte->relkind != RELKIND_VIEW) ||
-			 rte->tablesample || rte->inh == false) ||
-			((rte_other->relkind != RELKIND_RELATION && rte_other->relkind != RELKIND_VIEW) ||
-			 rte_other->tablesample || rte_other->inh == false) ||
-			(ts_is_hypertable(rte->relid) == ts_is_hypertable(rte_other->relid)))
+		if (rte->lateral || rte_other->lateral ||
+			(rte->relkind == RELKIND_VIEW && ts_is_hypertable(rte_other->relid)) ||
+			(rte_other->relkind == RELKIND_VIEW && ts_is_hypertable(rte->relid)) ||
+			(rte->relkind != RELKIND_VIEW && rte_other->relkind != RELKIND_VIEW &&
+			 (ts_is_hypertable(rte->relid) == ts_is_hypertable(rte_other->relid))))
 			ereport(ERROR,
 					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 					 errmsg("invalid continuous aggregate view"),
-					 errdetail("from clause can only have one hypertable and one normal table")));
+					 errdetail("views, lateral, multiple hypertables or normal tables"
+							   " are not supported in FROM clause")));
```

Comment:
This branch is a bit hard to follow... isn't better to split it in separated checks and with specific message and test case for each of them??

Reply:
Thanks for your review, I have broken it in multiple conditions. Regarding the test, they are already covered.